### ISSUE TITLE
Trouble loading GeoJSON files into Timeline

### ISSUE THREAD

**Issue by mattbeets** [skeate/Leaflet.timeline#37]
Hello people. 
I am doing a project, where I need to load some geoJSON files into a Leaflet map, shown in time with points and polygons. 
So far we had succes with loading and showing the GeoJSON files on the map:

```html
<!DOCTYPE html>
<html>
  <head>
      <script 
         src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.2.2/jquery.min.js">
      </script>
      
      <link rel="stylesheet" href="http://cdn.leafletjs.com/leaflet-0.7.3/leaflet.css" />
      
      <script src="http://cdn.leafletjs.com/leaflet-0.7.3/leaflet.js"></script>

      <script src="https://cdnjs.cloudflare.com/ajax/libs/moment.js/2.12.0/moment.min.js"></script>
      
      <!--<link rel="stylesheet" href="http://prehistoricmap.com/Timeline/dist/leaflet.timeline.min.css" /> -->
      
      <!--<script 
         src="http://prehistoricmap.com/Timeline/dist/leaflet.timeline.min.js">
      </script> -->
      
      
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width">
      
      <script>
var map;
function addDataToMap(data, map) {
    var dataLayer = L.geoJson(data);
    dataLayer.addTo(map);
}

$.getJSON("http://prehistoricmap.com/thanh_lab/latlong4.geojson", function(data) { addDataToMap(data, map); }); 

$(document).ready(function () {
  
map = L.map('map').setView([0, 0], 2);
  L.tileLayer('https://{s}.tiles.mapbox.com/v3/ebrelsford.ho06j5h0/{z}/{x}/{y}.png', {
      maxZoom: 18,
      attribution: 'Map data &copy; <a href="http://openstreetmap.org">OpenStreetMap</a> contributors, <a href="http://creativecommons.org/licenses/by-sa/2.0/">CC-BY-SA</a>, Imagery &copy; <a href="http://mapbox.com">Mapbox</a>'
      }).addTo(map);
 
 var slider = L.timelineSliderControl({
    formatOutput: function(date){
      return moment(date).format("YYYY");
    }
  });
  map.addControl(slider);

  var pointTimeline = L.timeline(points);
  pointTimeline.addTo(map);
  slider.addTimelines(pointTimeline);
});
    </script>

  </head>
  <body>
    <div id="map"></div>
  </body>
</html>
```

We also had success with showing the data in time with the Timeline (when copying the data directly into the HTML code):

```html
<!DOCTYPE html>
<html>
  <head>

    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.2.2/jquery.min.js"></script>

    <link rel="stylesheet" href="http://cdn.leafletjs.com/leaflet-0.7.3/leaflet.css" />
    <script src="http://cdn.leafletjs.com/leaflet-0.7.3/leaflet.js"></script>

    <script src="https://cdnjs.cloudflare.com/ajax/libs/moment.js/2.12.0/moment.min.js"></script>

    <script src="http://prehistoricmap.com/thanh_lab/Leaflet.timeline/dist/leaflet.timeline.min.js"></script>
    <link rel="stylesheet" href="http://prehistoricmap.com/thanh_lab/Leaflet.timeline/dist/leaflet.timeline.min.css"/>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width">

    <script>
var map;

var points = {
"type": "FeatureCollection",
"crs": { "type": "name", "properties": { "name": "urn:ogc:def:crs:OGC:1.3:CRS84" } },
"features": [
{ "type": "Feature", "properties": { "start": "1970", "end": "1972" }, "geometry": { "type": "Point", "coordinates": [ 11.924550479340329, 55.94619451008279 ] } },
{ "type": "Feature", "properties": { "start": "1975", "end": "1976" }, "geometry": { "type": "Point", "coordinates": [ 12.06219628503271, 55.909617190392566 ] } },
{ "type": "Feature", "properties": { "start": "1979", "end": "1980" }, "geometry": { "type": "Point", "coordinates": [ 12.06219628503271, 55.909617190392566 ] } },
{ "type": "Feature", "properties": { "start": "1980", "end": "1985" }, "geometry": { "type": "Point", "coordinates": [ 12.284822339303718, 55.639778377234506 ] } }
]
};

$(document).ready(function () {

  map = L.map('map').setView([0, 0], 2);

  L.tileLayer('https://{s}.tiles.mapbox.com/v3/ebrelsford.ho06j5h0/{z}/{x}/{y}.png', {
      maxZoom: 18,
      attribution: 'Map data &copy; <a href="http://openstreetmap.org">OpenStreetMap</a> contributors, <a href="http://creativecommons.org/licenses/by-sa/2.0/">CC-BY-SA</a>, Imagery &copy; <a href="http://mapbox.com">Mapbox</a>'
      }).addTo(map);

  var slider = L.timelineSliderControl({
    formatOutput: function(date){
      return moment(date).format("YYYY");
    }
  });
  map.addControl(slider);

  var pointTimeline = L.timeline(points);
  pointTimeline.addTo(map);
  slider.addTimelines(pointTimeline);
});
    </script>

  </head>
  <body>
    <div id="map"></div>
  </body>
</html>
```

The problem right now is combining these two things: 
Loading a GeoJSON file and showing it with the Timeline.

So far the code looks like this: 

```html
<!DOCTYPE html>
<html>
  <head>

    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.2.2/jquery.min.js"></script>

    <link rel="stylesheet" href="http://cdn.leafletjs.com/leaflet-0.7.3/leaflet.css" />
    <script src="http://cdn.leafletjs.com/leaflet-0.7.3/leaflet.js"></script>

    <script src="https://cdnjs.cloudflare.com/ajax/libs/moment.js/2.12.0/moment.min.js"></script>

    <script src="http://prehistoricmap.com/thanh_lab/Leaflet.timeline/dist/leaflet.timeline.min.js"></script>
    <link rel="stylesheet" href="http://prehistoricmap.com/thanh_lab/Leaflet.timeline/dist/leaflet.timeline.min.css"/>
    <script src="http://prehistoricmap.com/thanh_lab/leaflet.ajax.min.js"></script>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width">

    <script>
var map;

var points= new L.GeoJSON.AJAX("http://prehistoricmap.com/thanh_lab/clean4.geojson");


$(document).ready(function () {
 
  map = L.map('map').setView([0, 0], 2);

  L.tileLayer('https://{s}.tiles.mapbox.com/v3/ebrelsford.ho06j5h0/{z}/{x}/{y}.png', {
      maxZoom: 18,
      attribution: 'Map data &copy; <a href="http://openstreetmap.org">OpenStreetMap</a> contributors, <a href="http://creativecommons.org/licenses/by-sa/2.0/">CC-BY-SA</a>, Imagery &copy; <a href="http://mapbox.com">Mapbox</a>'
      }).addTo(map);

  var slider = L.timelineSliderControl({
    formatOutput: function(date){
      return moment(date).format("YYYY");
    }
  });
  map.addControl(slider);

  var pointTimeline = L.timeline(points);
  pointTimeline.addTo(map);
  slider.addTimelines(pointTimeline);
});
    </script>

  </head>
  <body>
    <div id="map"></div>
  </body>
</html>
```
 
This shows a map but no data. The Timeline is shown, but does not work. 
We have tried to take all the code from the GeoJSON and copy paste it into the exactly as it is in the file, and that works.
There must be a problem with the link between the GeoJSON file and making it load into the Timeline. 

We hope somebody can help.

**Comment by skeate** [comment]
This is probably way too late a response to help you (sorry), but it looks like the Leaflet.Ajax library actually creates a `L.GeoJSON` layer, which means it's incompatible with Leaflet.timeline.

The first example you gave is actually closer to what you'd want, only change the `addDataToMap` function:

```javascript
function addDataToMap(points, map) {
  var slider = L.timelineSliderControl({
    formatOutput: function(date){
      return moment(date).format("YYYY");
    }
  });
  map.addControl(slider);

  var pointTimeline = L.timeline(points);
  pointTimeline.addTo(map);
  slider.addTimelines(pointTimeline);
}
```

**Comment by mattbeets** [comment]
Hello, Skeate. 
Yes, we managed to combine the timeline and the GeoJSON layers using d3 instead.

But I am happy for a reply anyway :)
However, it seems to us that the timeline can only handle 2 GeoJSON layers at the same time. 
Is there any way to expand that to more than 2 layers in the Timeline code? Or are we just doing something wrong?

**Comment by skeate** [comment]
You should be able to add multiple timeline layers and control them all from a single `TimelineSliderControl`. Can you post your code?

**Comment by mattbeets** [comment]
<!DOCTYPE html>
<html>
  <head>
      <script 
         src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.2.2/jquery.min.js">
      </script>
      
      <link rel="stylesheet" href="http://cdn.leafletjs.com/leaflet-0.7.3/leaflet.css" />

      <script src="http://cdn.leafletjs.com/leaflet-0.7.3/leaflet.js"></script>

      <script src="https://d3js.org/d3.v3.min.js"></script> 
      
      <script src="https://cdnjs.cloudflare.com/ajax/libs/moment.js/2.12.0/moment.min.js"></script>
      
      <link rel="stylesheet" href="http://prehistoricmap.com/Timeline/dist/leaflet.timeline.min.css" /> 
      
      <script 
         src="http://prehistoricmap.com/thanh_lab/leaflet.timeline.min.js">
      </script>
      
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width">
      
      <script>d3.select(window).on("load", init)

       function init(){ 
   
        d3.json("cultures(2).geojson", function(polygons) { /*INDSÆT RIGTIG STI TIL FIL*/
          console.log(polygons);
          console.log(polygons[0]);      

      d3.json("international_textiles_time.geojson", function(points) { /*INDSÆT RIGTIG STI TIL FIL*/
          console.log(points);
          console.log(points[0]);
          
          /*Create the map*/
          map = L.map('map').setView([55, 51], 3);
          L.tileLayer('https://{s}.tiles.mapbox.com/v3/ebrelsford.ho06j5h0/{z}/{x}/{y}.png', {
              maxZoom: 18,
              attribution: 'Map data &copy; <a href="http://openstreetmap.org">OpenStreetMap</a> contributors, <a href="http://creativecommons.org/licenses/by-sa/2.0/">CC-BY-SA</a>, Imagery &copy; <a href="http://mapbox.com">Mapbox</a>'
              }).addTo(map);
                
            var slider = L.timelineSliderControl({
              formatOutput: function(date){
            return moment(date).format('Y');
                }
            });

          map.addControl(slider);
          
            var polygonTimeline = L.timeline(polygons, {
                 style: function(feature) {
                     return feature.properties.style;
                 },
                 onEachFeature: function(feature,layer) {
                         layer.bindPopup(feature.properties.Culture_na  /* put det her */);
                 }
             })
            polygonTimeline.addTo(map); 
      
            var pointTimeline = L.timeline(points,{
      pointToLayer: function(feature,latlng){
        var marker = L.marker(latlng);
        marker.bindPopup("<p><a title='link to publication' href="+ feature.properties.scientific + " target='_blank'><img src=" + feature.properties.photo + " height='80%' width='80%'></a></p>"
        + "<br><b>Find location:</b> " + feature.properties.Location_T
        + "<br><b>Associated culture:</b> " + feature.properties.Associated
        + "<br><b>Textile material:</b> " + feature.properties.Material
        + "<br><b>Comments:</b> " + feature.properties.Comments
        + "<br><b>Cal date (95,2% probability):</b> " + feature.properties.s2_95_star + " - " + feature.properties.s2_95_end
        + "<br><b>Archaeological site/context layer:</b> " + feature.properties.Site + "/" + feature.properties.Context_La
        + "<br><b>Museum collection:</b> " + feature.properties.Collection
        + "<br><b>Reference: </b>" + feature.properties.References
        );
        return marker;
      }
    });
            pointTimeline.addTo(map);
            

            
            slider.addTimelines(polygonTimeline,pointTimeline);
      });

         
                       

        }) /*End load Data*/


  }

    </script>

    <style>
/*
 * Everything fullscreen.
 */
html, body, #map {
  height: 100%;
  margin: 0;
  padding: 0;
  width: 100%;
}
    /*
     * You could style the timeline however you like, but we just make it wider here.
     */
    .leaflet-left {
      width: 100%;
    }
    </style>
  </head>
  <body>
    <div id="map"></div>
  </body>
</html>

**Comment by mattbeets** [comment]
At the moment there are only two layers loaded and this works, but as soon as we add one more layer in the same way, the page won't load.

**Comment by skeate** [comment]
I just tried using the above code (with different geojson files) and it worked. Can you post the code with all three?

**Comment by mattbeets** [comment]
Oops, my bad. 
It turns out I hadn't closed the funciton init(){ in the very end of the code. Now I have three working layers too! :D 
Those wretched { 's
Thanks so much for your help skeate!

**Comment by skeate** [comment]
Awesome!

I have found a bug around multiple `Timeline`s (they all share the same `times` array), so there might be some weird bugs (in particular if you try to remove some timelines), but an update is coming soon.

**Comment by mattbeets** [comment]
That sounds great!  Looking forward to the update.


Den 23. feb. 2017 5.37 PM skrev "Jonathan Skeate" <notifications@github.com
